[오렌지] 로또 2단계 미션 제출합니다.

YerinCho opened the pull request:
안녕하세요. 재택 중 외부 일이 생겨서 많이 늦었습니다. 죄송합니다.

리뷰 감사합니다!

YerinCho committed:
[Lotto] docs : 구현 목록 작성

YerinCho committed:
[Lotto] feat : LottoFactory

 로또 번호 생성 기능 추가
 무작위로 6개 숫자 선택해 로또 생성

YerinCho committed:
[Lotto] feat : LottoFactory

 로또 번호 생성 기능 추가
 무작위로 6개 숫자 선택해 로또 생성

YerinCho committed:
[Lotto] refactor : LottoFactory

 상수화, 불필요한 코드 삭제

YerinCho committed:
[Lotto] refactor : LottoFactory

 상수화, 불필요한 코드 삭제

YerinCho committed:
[Lotto] docs : 구현 목록 수정

YerinCho committed:
[Lotto] feat : PurchaseLotto

 로또 구매 갯수 계산

YerinCho committed:
[Lotto] feat : PurchaseLotto

 로또 구매 갯수 계산

YerinCho committed:
[Lotto] refactor : PurchaseLotto

 상수화

YerinCho committed:
[Lotto] refactor : PurchaseLotto

 상수화

YerinCho committed:
[Lotto] docs : 구현목록 수정

YerinCho committed:
[Lotto] refactor : LottoFactoryTest

 불필요한 코드 제거

YerinCho committed:
[Lotto] refactor : LottoFactoryTest

 불필요한 코드 제거

YerinCho committed:
[Lotto] feat : Lotto, 중복 검사 구현

YerinCho committed:
[Lotto] feat : Lotto, 중복 검사 구현

YerinCho committed:
[Lotto] feat : InputView

YerinCho committed:
[Lotto] docs : InputView

YerinCho committed:
[Lotto] feat : InputView 입력 예외처리

YerinCho committed:
[Lotto] feat : PurchaseLotto 입력 예외처리

 입력한 금액이 1000원 미만이면 예외처리

YerinCho committed:
[Lotto] feat : PurchaseLotto 입력 예외처리

 입력한 금액이 1000원 미만이면 예외처리

YerinCho committed:
[Lotto] docs : README 수정

 구입 금액 입력 예외처리
 보너스볼 입력 예외처리

YerinCho committed:
[Lotto] feat : Buyer 생성자 구현

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : Lotto.validateSize

 6자리인지 검사

YerinCho committed:
[Lotto] feat : Lotto.validateSize

 6자리인지 검사

YerinCho committed:
[Lotto] feat : SplitLottoNumbers

 , 기준으로 String 분리

YerinCho committed:
[Lotto] feat : Lotto.validateNumberScope

 로또 번호의 범위 유효성 검사 추가

YerinCho committed:
[Lotto] feat : Lotto.validateNumberScope

 로또 번호의 범위 유효성 검사 추가

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinningLotto

 보너스볼 범위 유효성 검사 추가
 Lotto.validateNumbersScope 분리

YerinCho committed:
[Lotto] feat : WinningLotto

 보너스볼 범위 유효성 검사 추가
 Lotto.validateNumbersScope 분리

YerinCho committed:
[Lotto] feat : WinningLotto 보너스볼 중복검사

 생성자 수정
 Lotto.lottoNumbers 접근제어자 변경

YerinCho committed:
[Lotto] feat : WinningLotto 보너스볼 중복검사

 생성자 수정
 Lotto.lottoNumbers 접근제어자 변경

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinnigResult 로또 일치 갯수 계산

YerinCho committed:
[Lotto] feat : WinningRankResult 보너스볼 일치 여부 확인

 WinnigResult -> WinningRankResult

YerinCho committed:
[Lotto] feat : WinningRankResult, WinningValue

 각 매수 별 당첨된 등수를 확인하는 기능 추가
 Lotto.getLottoNumbers 추가
 WinningLotto.isBonusNumber 추가
 WinningRankResult.isBonusNumberContain, WinningRankResult.checkCorrectNumber 의 파라미터 변경

YerinCho committed:
[Lotto] feat : RewardRate

 수익률 계산 기능 추가
 WinningRankResult.getWinningValueResult 추가

YerinCho committed:
[Lotto] refactor : RewardRate

 for 문 stream 으로 변경

YerinCho committed:
[Lotto] refactor : WinningRankResult.checkRank 메소드 분리

YerinCho committed:
[Lotto] refactor : Lotto.getLottoNumbers 불변성 보장

YerinCho committed:
`[Lotto] refactor : PurchaseLotto -> MoneyManager 클래스명 변경

 생성자 추가, 멤버변수 money 추가

YerinCho committed:
`[Lotto] refactor : PurchaseLotto -> MoneyManager 클래스명 변경

 생성자 추가, 멤버변수 money 추가

YerinCho committed:
[Lotto] refactor : Buyer 멤버변수 제거

YerinCho committed:
[Lotto] fix : InputView 입력 버퍼 초기화

YerinCho committed:
[Lotto] feat : Buyer.getLottos 추가

YerinCho committed:
[Lotto] feat : 초기 구매 화면 구현

 LottoMain 추가
 LottoManager 추가
 OutputView 추가

YerinCho committed:
[Lotto] feat : 당첨번호, 보너스번호 입력 구현

 LottoManager.setWinningLotto 추가

YerinCho committed:
[Lotto] feat : 당첨 결과 화면 구현

 LottoManager.analyzeLotto 추가
 WinningRankResult.winningValueResult 자료형 LinkedHashMap 으로 변경
 WinningValue 메시지 추가

YerinCho committed:
[Lotto] fix : 입력 버퍼 오류 수정

YerinCho committed:
[Lotto] feat : 수익률 계산 결과 출력 구현

 수익률 계산식 수정
 LottoManager.analyzeRewardRate 추가
 MoneyManager.getMoney 추가

YerinCho committed:
[Lotto] feat : 예외 메시지 추가

YerinCho committed:
[Lotto] refactor

 메소드, 변수명, 클래스명 변경
 메소드 분리
 접근 제한자 변경
 원시값 상수화
 메소드가 위치하는 클래스 변경
 불필요한 코드 정리
 README 수정

YerinCho committed:
[Lotto] feat : LottoNumber,ConvertInput 추가

 Integer 형식의 로또 넘버를 LottoNumber 로 포장
 그에 따른 ConvertInput 수정

YerinCho committed:
[Lotto] refactor : 패키지 분리

YerinCho committed:
[Lotto] feat : ConvertInputTest 공백 제거 검사

YerinCho committed:
[Lotto] feat : LottoNumber

 로또 숫자 범위 확인 테스트

YerinCho committed:
[Lotto] refactor : 상수화

YerinCho committed:
[Lotto] refactor : LottoFactory

 singleton 패턴으로 변경

YerinCho committed:
[Lotto] refactor : ConvertInput

 문자열 상수화

YerinCho committed:
[Lotto] refactor : LottoFactory

 singleton 패턴으로 변경

YerinCho committed:
[Lotto] refactor : LottoManager

 LottoManager 의 인스턴스 변수 Buyer, Money 제거
 main 으로 위치 변경

YerinCho committed:
[Lotto] refactor : LottoManager

 인스턴스 변수의 접근제어자 변경, getter 추가

YerinCho committed:
[Lotto] refactor : Money

 validateMoney 의 static 제거 및 그에 따른 테스트 변경

YerinCho committed:
[Lotto] refactor : LottoNumber

 validateNumberScope 의 static 제거 및 그에 따른 테스트 변경

YerinCho committed:
[Lotto] refactor : WinningValue, Money

 숫자에 _추가

YerinCho committed:
[Lotto] refactor : WinningValue

 메서드 분리

YerinCho committed:
[Lotto] refactor : LottoResult

 수익률 식 변경

YerinCho committed:
[Lotto] refactor : WinningLotto

 Lotto 를 상속하지 않고, Lotto WinningLotto 의 형식 사용으로 변경

YerinCho committed:
[Lotto] refactor : LottoResultTest

 수익률 식 변경

YerinCho committed:
[Lotto] refactor : LottoResultTest

 수익률 식 변경

YerinCho committed:
[Lotto] refactor : LottoManager, LottoAnalyzeManager

 LottoManager 에서 로또 결과 관련 메서드들을 LottoAnalyzeManager class 로 분리

YerinCho committed:
[Lotto] refactor : WinnigLotto

 보너스번호 관련 정보 추가

YerinCho committed:
[Lotto] refactor : WinnigLotto

 메서드 분리

YerinCho committed:
[Lotto] fix : LottoResult

 Int 오버플로우로 인한 식 수정

YerinCho committed:
[Lotto] fix : LottoResult

 Int 오버플로우로 인한 식 수정

YerinCho committed:
[Lotto] fix : LottoResult

 Int 오버플로우로 인한 식 수정

YerinCho committed:
[Lotto] refactor : WinningLotto

 isBonusNumber 에서, getLottoNumber를 사용하는 대신 Lotto에 isContainBonusNumber 메서드를 만들어 사용

YerinCho committed:
코드 초기화

YerinCho committed:
코드 초기화

YerinCho committed:
[Lotto - 수동] Update READEME

YerinCho committed:
[Lotto - 수동] Update READEME

YerinCho committed:
[Lotto] feat : LottoNumber

 로또의 각 숫자를 포장하는 클래스
 생성자 구현

YerinCho committed:
[Lotto] feat : LottoNumber

 로또의 각 숫자를 포장하는 클래스
 생성자 구현

YerinCho committed:
[Lotto] feat : LottoNumber

 로또 숫자의 유효성 검사 추가

YerinCho committed:
[Lotto] feat : LottoNumber

 로또 숫자의 유효성 검사 추가

YerinCho committed:
[Lotto] docs : README

 구현 기능목록 추가

YerinCho committed:
[Lotto] docs : README

 구현 기능목록 추가

YerinCho committed:
[Lotto] feat : LottoFactory

 싱글톤 패턴으로 LottoFactory 구현
 랜덤의 여섯자리 숫자를 반환하는 LottoFactory.createRandomLottoNumbers 생성

YerinCho committed:
[Lotto] feat : LottoFactory

 로또 수동 발행 기능 추가

YerinCho committed:
[Lotto] feat : LottoFactory

 로또의 중복을 막기 위해 List 대신 Set 자료구조로 변경

YerinCho committed:
[Lotto] feat : LottoFactory, LottoTicket

 로또 번호 개수 유효성 검사 (6개인지 확인)

YerinCho committed:
[Lotto] refactor : LottoFactory

 변수명 변경

YerinCho committed:
[Lotto] feat : WinningLotto, LottoTicket, LottoNumber

 보너스 숫자와 로또 번호의 유효성 검사 기능 추가
 LottoTicket.isContain 추가

YerinCho committed:
[Lotto] refactor : LottoTicket, WinningLotto

 상수의 접근제어자 변경

YerinCho committed:
[Lotto] feat : Spliter, InputView

 사용자 임의 번호 입력 기능 추가
 사용자 입력 , 기준으로 분리해 저장 기능 추가

YerinCho committed:
[Lotto] feat : Spliter, InputView

 사용자 임의 번호 입력 기능 추가
 사용자 입력 , 기준으로 분리해 저장 기능 추가

YerinCho committed:
[Lotto] refactor : Spliter

YerinCho committed:
[Lotto] refactor : Spliter

YerinCho committed:
[Lotto] feat : IssuedLottoTicket

 입력된 로또 발행 개수에 따라 로또 발행

YerinCho committed:
[Lotto] feat : IssuedLottoTicket

 입력된 로또 발행 개수에 따라 로또 발행

YerinCho committed:
[Lotto] feat : WinningValue

 결과값 저장하는 enum class 추가

YerinCho committed:
[Lotto] feat : WinningValue

 결과값 저장하는 enum class 추가

YerinCho committed:
[Lotto] feat : LottoTicket

 두 로또 비교해 맞은 개수 반환하는 기능 추가

YerinCho committed:
[Lotto] refactor : LottoTicket

 람다 이용해 인덴트 축소

YerinCho committed:
[Lotto] fix : WinningValue

 하나도 안 맞은 경우에 대해 값 수정

YerinCho committed:
[Lotto] fix : WinningValue

 하나도 안 맞은 경우에 대해 값 수정

YerinCho committed:
[Lotto] fix : WinningValue

 하나도 안 맞은 경우에 대해 값 수정

YerinCho committed:
[Lotto] fix : WinningValue

 하나도 안 맞은 경우에 대해 값 수정

YerinCho committed:
[Lotto] feat : WinningLotto

 보너스볼과 일치하는지 확인 기능 추가

YerinCho committed:
[Lotto] feat : LottoResult

 맞은 로또 번호 개수에 따라 등수 구하는 기능 추가

YerinCho committed:
[Lotto] feat : LottoResult

 맞은 로또 번호 개수에 따라 등수 구하는 기능 추가

YerinCho committed:
[Lotto] feat : LottoResult

 수익률 계산 추가

YerinCho committed:
[Lotto] feat : LottoResult

 수익률 계산 추가

YerinCho committed:
[Lotto] feat : Money

 Money 객체 추가 , 1000원 미만 예외처리

YerinCho committed:
[Lotto] feat : Money

 Money 객체 추가 , 1000원 미만 예외처리

YerinCho committed:
[Lotto] feat : InputView, OutputView

 입출력 추가

YerinCho committed:
[Lotto] feat : InputView, OutputView

 입출력 추가

YerinCho committed:
[Lotto] feat : LottoTicketCount

 구매한 티켓의 개수를 저장하는 객체 구현

YerinCho committed:
[Lotto] feat : LottoTicketCount

 구매한 티켓의 개수를 저장하는 객체 구현

YerinCho committed:
[Lotto] fix : IssuedLottoTicket

 LottoTicketCount 클래스 구현에 따른 변경사항 적용

YerinCho committed:
[Lotto] fix : IssuedLottoTicket

 LottoTicketCount 클래스 구현에 따른 변경사항 적용

YerinCho committed:
[Lotto] feat : LottoMain

 main 구현

YerinCho committed:
[Lotto] feat : LottoMain

 main 구현

YerinCho committed:
[Lotto] refactor

 getter 생성
 상수화
 출력내용 수정

YerinCho committed:
[Lotto] refactor

 상수화
 접근제어자 변경
 null 체크 추가

YerinCho committed:
[Lotto] feat : LottoController

 Main 에서 Controller 분리

YerinCho committed:
merge

YerinCho committed:
merge

YerinCho committed:
초기화

YerinCho committed:
merge

jeonghoon1107 commented on src/main/java/lotto/Controller/LottoController.java:
```diff
@@ -0,0 +1,40 @@
+import lotto.view.OutputView;
+
+public class LottoController {
+    private Money money;
```
인스턴스 변수가 너무 많은데 메서드에서 값을 반환하는 식으로 해서 변수를 줄여보면 어떨까요?

YerinCho replied:
buyLotto에서 ticketCount 자체를 반환하게 해서 인스턴스 변수를 하나 줄이고, 한 메소드 내에서만 사용되는 변수는 메소드 내에서만 호출되도록 변경했습니다.

jeonghoon1107 commented on src/main/java/lotto/domain/LottoFactory.java:
```diff
@@ -25,12 +23,22 @@ public static LottoFactory getInstance() {
         return lotto;
     }
+
+    public static Set<LottoNumber> createManualLottoNumbers(List<Integer> manualLottoNumbers) {
```
외부에서 자동, 수동 로또 만든 것을 합치지 않고 해당 클래스에서 하면 어떨까요?
한번에 `Set<LottoNumber>` 을 return할 수 있도록.

또한 자동, 수동 로또를 생성하는 행위를 interface로 추출하면 역할대로 분리할 수 있을 것 같아요.

```
public interface LottoGenerator {
   Set<LottoNumber>generate(dto);
}
```

YerinCho replied:
생각보다 어렵네요,, 제가 한 방식은 LottoGenerater 인터페이스를 만들고, 새로 자동, 수동 로또를 생성하는 클래스를 따로 만들어 구현했습니다! DTO 방식은 아직 이해가 잘 안가서 적용하진 않았습니다.  :(

jeonghoon1107 reviewed (changes requested):
안녕하세요 오렌지!
아쉬운 점이 있어서 피드백 남겼으니 확인해주세요 :)

YerinCho committed:
[Lotto] refactor : LottoController

 인스턴스 변수 간소화
 충돌 해결 과정에서 생겼던 클래스 제거

YerinCho committed:
[Lotto] refactor : IssuedLottoTicket, LottoFactory

 LottoFactory 에서 수동, 자동 통합된 Set<LottoTicket>을 리턴하도록 변경
 충돌 해결 과정에서 생겼던 클래스 제거

YerinCho committed:
[Lotto] refactor : LottoGenerator, ManualLottoGenerator, RandomLottoGenerator

 자동, 수동 번호 생성하는 LottoGenerator Interface 구현
 LottoGenerator 의 구현에 따른 ManualLottoGenerator, RandomLottoGenerator 구현
 LottoFactoryTest 의 테스트 코드를 ManualLottoGeneratorTest 와 RandomLottoGeneratorTest 두 곳으로 분리
 변경사항에 맞게 InputView, IssuedLottoTicket 수정

jeonghoon1107 reviewed (approved):
안녕하세요 오렌지!
리팩토링 깔끔하네요 👍 
미션 진행하느라 고생하셨습니다 :)
